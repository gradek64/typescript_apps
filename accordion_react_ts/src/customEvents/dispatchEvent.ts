const customEvent = () => {
  //this needs to be assign to regular function because we want to have access to this;
  const events = {};
  // const on = (id, fn) => {}
  type Callback = (payload?: unknown) => any;
  const on = function (id: string, fn: Callback) {
    if (id && !this.events[id]) {
      this.events[id] = [];
    }
    if (!id || typeof fn !== 'function') {
      console.warn('you event callback has to be function or id doesnt exists');
      return;
    }
    this.events[id].push(fn);
  };
  const emit = function (id: string, payload?: unknown) {
    if (!Array.isArray(this.events[id])) return;
    this.events[id].forEach((callback: Callback) => callback(payload));
  };

  return {
    on,
    emit,
    events,
  };
};

export default customEvent();
