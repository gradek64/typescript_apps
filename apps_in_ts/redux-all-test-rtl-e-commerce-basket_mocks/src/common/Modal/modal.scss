.clickAway {
  width: 100vw;
  height: 100vh;
  position: fixed;
  display: flex;
  top: 0;
  left: 0;
}

.modal {
  width: 100vw;
  height: 100vh;
  position: fixed;
  display: flex;
  align-items: center;
  justify-content: center;
  top: 0;
  left: 0;
  background-color: rgba(0, 0, 0, 0.5);
  z-index: 100;

  .modal-content {
    padding: 2rem;
    width: 80%;
    max-width: 1000px;
    background-color: white;
    position: relative;
  }

  .close-modal {
    position: absolute;
    max-width: 100px;
    width: 5%;
    height: auto;
    top: -1rem;
    border-radius: 50%;
    border: 1px solid;
    cursor: pointer;
    right: -1rem;
    font-size: 3rem;
  }
}
