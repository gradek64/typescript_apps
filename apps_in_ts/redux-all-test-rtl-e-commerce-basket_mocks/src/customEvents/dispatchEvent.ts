const customEvent = () => {
  const events:{[key:string]:any} = {};
  const on = (id:string, fn:Function) => {
    if (id && !events[id]) {
       events[id] = [];
    }
    events[id].push(fn);
  };

  const emit =  (id:string, payload?:any) => {
    if (!Array.isArray(events[id])) return;
     events[id].forEach((callback:Function) => callback(payload));
  };

  return {
    on,
    emit,
    events,
  };
};

export default customEvent();
