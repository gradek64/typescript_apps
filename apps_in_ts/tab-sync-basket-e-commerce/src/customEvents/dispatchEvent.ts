const customEvent = () => {
  type eventWillReturn = Function[]|null
  const events:{[key:string]:eventWillReturn} = {};
  const on = (id:string, fn:Function) => {
    if (id && !events[id]) {
       events[id] = [];
    }
    events[id]!.push(fn);
  };

  const cancel = (id:string) => {
    if (!Array.isArray(events[id])) return;
    events[id] = null;
  }

  const emit =  (id:string, payload?:any) => {
    if (!Array.isArray(events[id])) return;
     events[id]!.forEach((callback:Function) => callback(payload));
  };

  return {
    on,
    emit,
    cancel,
    events,
  };
};

export default customEvent();
