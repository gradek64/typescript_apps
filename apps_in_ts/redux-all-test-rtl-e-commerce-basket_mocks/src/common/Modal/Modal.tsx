import React, { useState, useEffect } from 'react';

// custom
import dispatchEvent from '../../customEvents/dispatchEvent';
import './modal.scss';

interface IProps {
  eventOn: string,
  children: React.ReactNode
}

const Modal = ({ eventOn, children }: IProps) => {

  const [open, openState] = useState(false);
  const [didMount, setDidMount] = useState(false);

  useEffect(() => {
    setDidMount(true);
    return () => setDidMount(false);
  }, [])

  if (!didMount) {
    return null;
  }


  const openModal = () => {
    const delaySelector = setTimeout(() => {
      clearTimeout(delaySelector);
      (document.querySelector('[tabIndex="-1"]') as HTMLElement)?.focus();
    }, 100);
    openState(true);
  };
  const closeModal = (evt: React.MouseEvent | React.KeyboardEvent) => {
    if (evt && evt.target !== evt.currentTarget) return;
    if (evt.type === 'keydown') {
      const { key } = evt as React.KeyboardEvent
      if (key !== 'Escape') return;
    }
    openState(false);
  };

  //register event with custom event
  dispatchEvent.on(eventOn, openModal);

  return (
    <React.Fragment>
      <div
        style={open ? { display: 'flex' } : { display: 'none' }}
        tabIndex={open ? -1 : 0}
        role="presentation"
        className="modal"
        onClick={closeModal}
        onKeyDown={closeModal}
      >
        <div className="modal-content">{children}</div>
      </div>
    </React.Fragment>
  );
};


export default Modal;
