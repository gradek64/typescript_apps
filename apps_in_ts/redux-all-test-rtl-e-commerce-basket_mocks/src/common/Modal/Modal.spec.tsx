import React from 'react';
import { render, fireEvent, cleanup, act } from '@testing-library/react'
import Modal from './Modal';
import dispatchEvent from '../../customEvents/dispatchEvent'

const reactNode = () => (<p>modal content</p>)
const props = {
  children: reactNode(),
  eventOn: 'OPEN_MODAL',
}

const getElements = (container: HTMLElement) => {
  const modalBody = container.querySelector('.modal')
  const modalContent = container.querySelector('.modal-content')
  return { modalBody, modalContent }
}

// updates hook needs to be wrapped with act
const openModalWithCustomEvent = () => act(() => {
  dispatchEvent.emit('OPEN_MODAL')
});

afterEach(cleanup)
describe('render', () => {
  it('opens a modal', () => {
    const { asFragment } = render(<Modal {...props} />);
    openModalWithCustomEvent()
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('events', () => {
  describe('click on modal content and modal body', () => {
    it('should close modal when click outside the content', () => {
      const { container } = render(<Modal {...props} />);
      openModalWithCustomEvent()
      const { modalBody } = getElements(container)
      fireEvent.click(modalBody as HTMLElement)
      const computedStyle = window.getComputedStyle(modalBody!, null);
      const { _values } = computedStyle as any
      expect(_values.display).toBe('none')
    })
    it('should NOT close modal when click inside the modal content', () => {
      const { container } = render(<Modal {...props} />);
      openModalWithCustomEvent()
      const { modalBody, modalContent } = getElements(container)

      fireEvent.click(modalContent as HTMLElement)
      const computedStyle = window.getComputedStyle(modalBody!, null);
      const { _values } = computedStyle as any
      expect(_values.display).toBe('flex')
    })
  })
  describe('press on ESC key', () => {
    it('should close modal when ESCAPE key press', () => {
      const { container } = render(<Modal {...props} />);
      const { modalBody } = getElements(container)
      openModalWithCustomEvent()
      fireEvent.keyDown(modalBody as HTMLElement, { key: 'Escape' })
      const computedStyle = window.getComputedStyle(modalBody!, null);
      const { _values } = computedStyle as any
      expect(_values.display).toBe('none')
    })
    it('should NOT close modal when any other key press but ESCAPE', () => {
      const { container } = render(<Modal {...props} />);
      const { modalBody } = getElements(container)
      openModalWithCustomEvent()
      fireEvent.keyDown(modalBody as HTMLElement, { key: 'k' })
      const computedStyle = window.getComputedStyle(modalBody!, null);
      const { _values } = computedStyle as any
      expect(_values.display).toBe('flex')
    })
  })
})
