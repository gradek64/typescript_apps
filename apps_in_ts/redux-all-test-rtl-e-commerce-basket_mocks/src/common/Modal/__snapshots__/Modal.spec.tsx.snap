// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render opens a modal 1`] = `
<DocumentFragment>
  <div
    class="modal"
    role="presentation"
    style="display: flex;"
    tabindex="-1"
  >
    <div
      class="modal-content"
    >
      <p>
        modal content
      </p>
    </div>
  </div>
</DocumentFragment>
`;
